lib/nerves_hub_link/features/features.ex

```diff
@@ -0,0 +1,229 @@
+          {feature, %{module: ^module}} <- state.features,
+          features == :all or feature in features,
+          reduce: state do
+        acc ->
+          # Ignore since either :ok, or {:error, :not_found}
+          _ = DynamicSupervisor.terminate_child(FeaturesSupervisor, pid)
+          _ = Socket.push("features", "#{feature}:detached", %{})
+          put_in(acc.features[feature][:attached?], false)
+      end
+
+    {:noreply, state}
+  end
+
+  def handle_cast({:attach, features}, state) do
+    features = if features == :all, do: Map.keys(state.features), else: features
+
+    state =
+      for feature <- features, reduce: state do
+        acc ->
+          with mod when not is_nil(mod) <- state.features[feature][:module],
+               :ok <- start_feature(mod) do
+            _ = Socket.push("features", "#{feature}:attached", %{})
+            put_in(acc.features[feature][:attached?], true)
+          else
+            error ->
+              reason = if error, do: "start_failure", else: "unknown_feature"
+              _ = Socket.push("features", "#{feature}:error", %{reason: reason})
+
+              Logger.warning(
+                "[NervesHubLink.Features] failed to start #{feature}: #{inspect(error)}"
+              )
+
+              acc
+          end
+      end
+
+    {:noreply, state}
+  end
+
+  def handle_cast({:handle_event, event, payload}, state) when event in ["attach", "detach"] do
+    features =
+      case payload["features"] do
+        "all" ->
+          :all
+
+        features when is_list(features) ->
+          features
+
+        feat when is_binary(feat) ->
+          [feat]
+
+        unknown ->
+          Logger.warning(
+            "[NervesHubLink.Features] missing features to #{event}: Got #{inspect(unknown)}"
+          )
+
+          []
+      end
+
+    handle_cast({String.to_existing_atom(event), features}, state)
+  end
+
+  def handle_cast({:handle_event, event, payload}, state) do
+    results =
+      for {feature, %{module: module}} <- state.features do
+        case event do
+          <<^feature::binary, ":", event::binary>> ->
+            send(module, {:__feature_event__, event, payload})
+
+          _ ->
+            nil
+        end
+      end
+
+    if results == [] do
+      # Event was unhandled. Maybe report it to NH?
+      Logger.warning(
+        "[NervesHubLink.Features] Unhandled event: #{inspect(event)} - #{inspect(payload)}"
+      )
+    end
+
+    {:noreply, state}
+  end
+
+  defp start_feature(feature_module) do
+    result = DynamicSupervisor.start_child(FeaturesSupervisor, feature_module)
+
+    case result do
+      {:ok, _pid} -> :ok
+      {:ok, _pid, _info} -> :ok
+      {:error, {:already_started, _pid}} -> :ok
+      error -> error
+    end
+  end
+
+  defmacro __using__(opts) do
+    name = opts[:name] || raise "Missing required feature arg: name"
+    version = opts[:version] || raise "Missing required feature arg: version"
```

Comment:
Changed to being arguments to the `use` macro rather than callbacks so that it can be enforced at compile time vs just a warning (not everyone has `--warnings-as-errors` flag by default)

These need some typing and docs that will come later as I'm out of time for a bit